File: src/main/java/servlet/Assignment5.java
package servlet;

import java.io.PrintWriter;
import java.io.IOException;
import java.util.*;

import javax.servlet.ServletException;
import javax.servlet.ServletOutputStream;
import javax.servlet.annotation.WebServlet;
import javax.servlet.http.HttpServlet;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

@WebServlet(
        name = "/assignment5",
        urlPatterns = {"/assignment5"}
)

public class Assignment5 extends HttpServlet {

    // Other strings
    static String Style = "https://mason.gmu.edu/~dparikh4/assignment5/assignment5.css";
    static String topBodyHTML = "<body>\n" +
            "        <script src=\"https://mason.gmu.edu/~dparikh4/assignment5/assignment5.js\"></script>\n" +
            "\n" +
            "        <!-- Title Area -->\n" +
            "        <div class=\"main-section\">\n" +
            "                <div class=\"a3-subsection\">\n" +
            "                        <h1>\n" +
            "                                SWE432-001 Assignment 5\n" +
            "                        </h1>\n" +
            "                        <h3>\n" +
            "                                Created By:\n" +
            "                                <a href=\"https://mason.gmu.edu/~afuentel/\">\n" +
            "                                        Alexandro Fuentelzas\n" +
            "                                </a>\n" +
            "                                ,&nbsp;\n" +
            "                                <a href=\"https://mason.gmu.edu/~dparikh4/\">\n" +
            "                                        Deval Parikh\n" +
            "                                </a>\n" +
            "                                ,&nbsp;\n" +
            "                                <a href=\"https://mason.gmu.edu/~spolina/\">\n" +
            "                                        Surya Polina\n" +
            "                                </a>\n" +
            "                        </h3>\n" +
            "                </div>\n" +
            "\n" +
            "                <!-- Forms and User Area -->\n" +
            "                <div class=\"a3-subsection\">\n" +
            "\n" +
            "\n" +
            "                        <!-- Variables Table -->\n" +
            "                        <h1>Define Variables</h1>\n" +
            "                        <h3>Define your variables here to be used in the logical predicates below. Type a variable name and click the plus button to add a new variable.</h3>\n" +
            "                        <form class=\"form\" name=\"form0\" method=\"post\" action=\"\">\n" +
            "                                <table class=\"table-form\" id=\"var_table\">\n" +
            "                                        <tbody><tr>\n" +
            "                                                <th>Variable</th>\n" +

            "\n" +
            "                                                <th><button type=\"button\" class=\"fa fa-plus button\" style=\"font-size:24px\" onclick=\"addVarRow()\"></button></th>\n" +
            "                                        </tr>\n" +
            "                                        <tr onmouseover=\"var_table.clickedRowIndex=this.rowIndex\">\n" +
            "                                                <td><input type=\"text\" name=\"defined_variable_1[]\" id=\"defined_variable_1\"></td>\n" +
            "                                        </tr>\n" +
            "\n" +
            "                                </tbody></table>\n" +
            "                                <p class=\"error-text\" id=\"varErrorMessage\"></p>\n" +
            "\n" +
            "\n" +
            "                        </form>\n" +
            "\n" +
            "                        <!-- Predicates Table -->\n" +
            "                        <h1>Logical Predicates</h1>\n" +
            "                        <h3>Create your logical predicates here. For boolean variables, you can either choose a default boolean value or select from your defined variables. The second logical operator is to link together the next logical predicate. Click the plus button to add more logical predicates. Once you are finished adding predicates, click the Submit button to obtain your results.</h3>\n" +
            "                        <form class=\"form\" method=\"post\" action=\"./assignment5\">\n" +
            "                                <table class=\"table-form\" id=\"logic_table\">\n" +
            "                                        <tbody><tr>\n" +
            "                                                <th>Boolean Variable 1</th>\n" +
            "                                                <th>Logical Operator</th>\n" +
            "                                                <th>Boolean Variable 2</th>\n" +
            "                                                <th>Logical Operator</th>\n" +
            "                                                <th><button type=\"button\" class=\"fa fa-plus button\" style=\"font-size:24px\" onclick=\"addPredRow()\"></button></th>\n" +
            "                                        </tr>\n" +
            "                                        <tr onmouseover=\"logic_table.clickedRowIndex=this.rowIndex\">\n" +
            "                                                <td>\n" +
            "                                                        <select type=\"text\" name=\"Boolean1_1[]\" id=\"ops1_1\">\n" +
            "                                                                <option value=\"true\">True</option>\n" +
            "                                                                <option value=\"false\">False</option>\n" +
            "                                                        </select>\n" +
            "                                                </td>\n" +
            "                                                <td>\n" +
            "                                                        <select type=\"text\" name=\"LogicOp1_1[]\" id=\"LogicOp1_1\">\n" +
            "                                                                <optgroup label=\"Logical Operators\">\n" +
            "                                                                        <option value=\"logic-and\">&amp;&amp; - AND</option>\n" +
            "                                                                        <option value=\"logic-or\">|| - OR</option>\n" +
            "                                                                        <option value=\"logic-xor\">^ - XOR</option>\n" +
            "                                                                </optgroup>\n" +
            "                                                                <optgroup label=\"Bitwise Operators\">\n" +
            "                                                                        <option value=\"bitwise-and\">&amp; - AND</option>\n" +
            "                                                                        <option value=\"bitwise-or\">| - OR</option>\n" +
            "                                                                        <option value=\"bitwise-xor\">^ - XOR</option>\n" +
            "                                                                </optgroup>\n" +
            "                                                        </select>\n" +
            "\n" +
            "                                                </td>\n" +
            "                                                <td>\n" +
            "                                                        <select type=\"text\" name=\"Boolean2_1[]\" id=\"ops2_1\">\n" +
            "                                                                <option value=\"true\">True</option>\n" +
            "                                                                <option value=\"false\">False</option>\n" +
            "                                                        </select>\n" +
            "                                                </td>\n" +
            "                                                <td>\n" +
            "                                                        <select type=\"text\" name=\"LogicOp2_1[]\" id=\"LogicOp2_1\">\n" +
            "                                                                <option value=\"none\">None</option>\n" +
            "                                                                \n" +
            "                                                                <optgroup label=\"Logical Operators\">\n" +
            "                                                                        <option value=\"logic-and\">&amp;&amp; - AND</option>\n" +
            "                                                                        <option value=\"logic-or\">|| - OR</option>\n" +
            "                                                                        <option value=\"logic-xor\">^ - XOR</option>\n" +
            "                                                                </optgroup>\n" +
            "                                                                <optgroup label=\"Bitwise Operators\">\n" +
            "                                                                        <option value=\"bitwise-and\">&amp; - AND</option>\n" +
            "                                                                        <option value=\"bitwise-or\">| - OR</option>\n" +
            "                                                                        <option value=\"bitwise-xor\">^ - XOR</option>\n" +
            "                                                                </optgroup>\n" +
            "                                                        </select>\n" +
            "\n" +
            "                                                </td>\n" +
            "                                                <td><button type=\"button\" class=\"fa fa-trash button\" style=\"font-size:24px\" onclick=\"delPredRow()\"></button></td>\n" +
            "                                        </tr>\n" +
            "\n" +
            "                                </tbody></table>\n" +
            "                                <br>\n" +
            "                                <input class=\"submit-button button\" type=\"submit\" name=\"Submit\" value=\"Submit\">\n" +
            "                        </form>\n";

    static String endBodyHTML = "\n" +
            "                        <h1>Collaboration Summary</h1>\n" +
            "                        <h3>(UPDATE FOR ASSIGNMENT 5) Our group collaboration to complete this assignment. We all contributed to each other's parts. Deval contributed by creating Javascript functions to add and remove logical predicates. Alex contributed by creating Javascript functions to add and remove variables. Surya worked on HTML and CSS of the page. We collaborated together to design the overall usability, layout, and format of the user interface. Alex came up with the variable definition design structure. Surya performed usability testing to help create helpful instructions. Deval wrote instructions.\n" +
            "                        </h3>\n" +
            "                </div>\n" +
            "        </div>\n";


    public void doPost(HttpServletRequest request, HttpServletResponse response)
            throws ServletException, IOException {

        // HTTP POST request backend logic

        response.setContentType("text/html");
        PrintWriter out = response.getWriter();

        Map formData = request.getParameterMap();

        PrintHead(out);
        PrintPostBody(out, formData);
        PrintTail(out);
    }

    @Override
    protected void doGet(HttpServletRequest req, HttpServletResponse res)
            throws ServletException, IOException {

        res.setContentType("text/html");
        PrintWriter out = res.getWriter();

        PrintHead(out);
        PrintBody(out);
        PrintTail(out);

        out.flush();
        out.close();

    }



    // HTML functions for doGet



    private void PrintHead(PrintWriter out) {
        out.println("<html>");
        out.println("");

        out.println("<head>");
        out.println("<title>Assignment 5</title>");
        // Import CSS
        out.println("<link rel=\"stylesheet\" type=\"text/css\" href=\"" + Style + "\">");
        out.println("<link rel=\"stylesheet\" href=\"https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css\">");
        out.println("</head>");
        out.println("");
    }

    private void PrintBody(PrintWriter out) {
        out.println(topBodyHTML);
        out.println(endBodyHTML);
    }

    private void PrintTail(PrintWriter out) {
        out.println("");
        out.println("</html>");
    }



    // HTML functions for doPost


    // HTML Body from doPost
    private void PrintPostBody(PrintWriter out, Map formData) {
        out.println(topBodyHTML);
        PrintUserPredicates(out, formData);
        out.println(endBodyHTML);
    }

    // Helper to generate HTML for user inputted predicates
    private void PrintUserPredicates(PrintWriter out, Map formData) {
        String predicateOutputHTML = "\n" +
                "                        <h1>Final Predicate</h1>\n" +
                "                        <h3>This is your inputted predicate put together! We are using the symbol" +
                " notation for logical operators for clarity and consistency because it makes no difference in regards to a truth table.</h3>\n";
        out.println(predicateOutputHTML);
        out.println("<div class=\"predicates\">");


        Set s = formData.entrySet();
        Iterator it = s.iterator();
        int count = 1;

        while(it.hasNext()){

            Map.Entry<String,String[]> entry = (Map.Entry<String,String[]>)it.next();

            String key             = entry.getKey();
            String[] value         = entry.getValue();

//            out.println("Key is "+key+"<br>");
//            if(value.length>1){
//                out.println("test");
//                for (int i = 0; i < value.length; i++) {
//                    out.println("<li>" + value[i].toString() + "</li><br>");
//                }
//            }else
//                out.println("Value is "+value[0].toString()+"<br>");


            if(key.contains("Logic")) {
                String cur_op = value[0];
                // Check if last element and a logical operator selected (formData.size()-1 to ignore the submit button)
                if(!cur_op.equals("none") && count % 4 == 0 && formData.size()-1 == count) {
                    // Ignore selected logical operator
                    continue;
                }
                if(cur_op.contains("and")) {
                    out.println("& ");
                } else if(cur_op.contains("xor")) {
                    out.println("^ ");
                } else if(cur_op.contains("or")) {
                    out.println("| ");
                } else {
                    if(count % 4 != 0) {
                        out.println("ERROR");
                    }
                }

            } else if(key.contains("Bool")) {
                out.println(value[0]+" ");
            }


            count++;
        }
        out.println("</div>");

    }


}

File: src/main/java/servlet/assignment5.js

///////////////////
//    Globals    //
///////////////////

var recordedVariableNames = new Set(); // Already used variable names
var varCount = 1; // Used to set unique IDs for variables
var predCount = 1; // Used to set unique IDs for predicates

////////////////////////////////////////////
//    Define Variables Table Functions    //
////////////////////////////////////////////

// Adds a new variable to the table of defined variables
function addVarRow() {

    var userInputtedVariableName = document.getElementById('defined_variable_1').value;
    var userInputtedVariableValue = document.getElementById('defined_value_1').value;

    // Check if variable name already exists
    if(recordedVariableNames.has(userInputtedVariableName)) {
        console.log(userInputtedVariableName);
        document.getElementById("varErrorMessage").innerHTML = 'Variable name "' + userInputtedVariableName + '" already exists. Please enter a unique variable name.';
        return;
    } else {
        recordedVariableNames.add(userInputtedVariableName);
        document.getElementById("varErrorMessage").innerHTML = ''

        varCount++;
        var oRow = document.getElementById("var_table").insertRow(document.getElementById("var_table").rows.length);
        oRow.onmouseover = function () { document.getElementById("var_table").clickedRowIndex = this.rowIndex; };
        var oCell1 = oRow.insertCell(0);
        var oCell2 = oRow.insertCell(1);
        var oCell3 = oRow.insertCell(2);
        var oCell4 = oRow.insertCell(3);

        var curRowIndex = document.getElementById("var_table").rows.length - 1;

        oCell1.innerHTML = "<p id='defined_variable_" + varCount + "' >" + userInputtedVariableName + "</p>";
        oCell2.innerHTML = "=";
        oCell3.innerHTML = "<p id='defined_value_" + varCount + "' >" + userInputtedVariableValue + "</p>";
        oCell4.innerHTML = "<th><button type=button class='fa fa-trash button' style='font-size:24px' onClick='delVarRow()'></button></th>"



        // Go through each logical predicate row and append to boolean variable options
        var logicTable = document.getElementById('logic_table');
        var logicRows = logicTable.rows;
        for(var r = 1, n = logicRows.length; r < n; r++) {

            var curLogicRow = logicRows[r];
            console.log(curLogicRow.cells[0]);

            // Extract current logical predicate row's ID
            var curRow = curLogicRow.cells[0].innerHTML;
            var curRowID = curRow.match('["].\+["]')[0];
            var curRowIDNum = curRowID.match('[_][0-9]+')[0].substring(1);

            console.log('ops1_' + curRowIDNum);

            var boolvar1 = document.getElementById('ops1_' + curRowIDNum);
            var boolvar2 = document.getElementById('ops2_' + curRowIDNum);

            var newOption1 = document.createElement("option");
            newOption1.text = document.getElementById("defined_variable_1").value;
            newOption1.value = document.getElementById('defined_value_1').value;

            var newOption2 = document.createElement("option");
            newOption2.text = document.getElementById("defined_variable_1").value;
            newOption2.value = document.getElementById('defined_value_1').value;

            boolvar1.add(newOption1);
            boolvar2.add(newOption2);

        }

        document.getElementById("defined_variable_1").value = "";
    }


}

// Deletes defined variable
function delVarRow() {

    // Delete from recoreded variable names
    var selectedIndex = document.getElementById("var_table").clickedRowIndex;
    var selectedVariable = document.getElementById("var_table").rows[selectedIndex].cells[0].innerHTML;
    var extractedText = selectedVariable.match("[>].\+[<]")[0];
    extractedText = extractedText.substring(1, extractedText.length-1);
    recordedVariableNames.delete(extractedText);

    // Delete from list of variables
    document.getElementById("var_table").deleteRow(selectedIndex);

    // Delete from options drop down in predicates table
    var logicTable = document.getElementById('logic_table');
    var logicRows = logicTable.rows;
    for (var r = 1, n = logicRows.length; r < n; r++) {

        // Extract current variable row's ID
        var curRow = logicRows[r].cells[0].innerHTML;
        var curRowID = curRow.match('["].\+["]')[0];
        var curRowIDNum = curRowID.match('[_][0-9]+')[0].substring(1);

        var boolvar1 = document.getElementById('ops1_' + curRowIDNum);
        var boolvar2 = document.getElementById('ops2_' + curRowIDNum);
        boolvar1.remove(document.getElementById("var_table").clickedRowIndex);
        boolvar2.remove(document.getElementById("var_table").clickedRowIndex);
    }
}

//////////////////////////////////////////////
//    Logical Predicates Table Functions    //
//////////////////////////////////////////////

// Adds a new logical predicate statement to the predicate table
function addPredRow() {
    predCount++;
    var oRow = document.getElementById("logic_table").insertRow(document.getElementById("logic_table").rows.length);
    oRow.onmouseover = function () { document.getElementById("logic_table").clickedRowIndex = this.rowIndex; };
    var oCell1 = oRow.insertCell(0);
    var oCell2 = oRow.insertCell(1);
    var oCell3 = oRow.insertCell(2);
    var oCell4 = oRow.insertCell(3);
    var oCell5 = oRow.insertCell(4);


    var curRowIndex = document.getElementById("logic_table").rows.length - 1;
    console.log(curRowIndex)

    // Go through each defined variable row and populate the boolean variable options

    var varTable = document.getElementById('var_table');

    var bool1Options = `<select type=text name=Boolean1_` + predCount + `[] id="ops1_` + predCount + `">`;
    var bool2Options = `<select type=text name=Boolean2_` + predCount + `[] id="ops2_` + predCount + `">`;

    var defaultOptions = `<option value="true">True</option><option value="false">False</option>`;
    bool1Options += defaultOptions;
    bool2Options += defaultOptions;


    // Instead of normal for loop, go through elements from DOM
    var varRows = varTable.rows
    for(var r = 2, n = varRows.length; r < n; r++) {

        // Extract current variable row's ID
        var curRow = varRows[r].cells[0].innerHTML;
        var curRowID = curRow.match('["].\+["]')[0];
        var curRowIDNum = curRowID.match('[_][0-9]+')[0].substring(1);

        console.log(curRowID);
        console.log(curRowIDNum);

        var curVariable = document.getElementById('defined_variable_' + curRowIDNum).innerHTML;
        var curValue = document.getElementById('defined_value_' + curRowIDNum).innerHTML;
        console.log(curVariable);
        bool1Options += `<option value="` + curValue + `">` + curVariable + `</option>`;
        bool2Options += `<option value="` + curValue + `">` + curVariable + `</option>`;

    }

    bool1Options += `</select>`;
    bool2Options += `</select>`;

    oCell1.innerHTML = bool1Options;

    oCell2.innerHTML = `
        <select type=text name=LogicOp1_` + predCount + `[] id=LogicOp1_` + predCount + `>                                       
            <optgroup label="Logical Operators">
                    <option value="logic-and">&&</option>
                    <option value="logic-or">||</option>
            </optgroup>
            <optgroup label="Bitwise Operators">
                    <option value="bitwise-and">&</option>
                    <option value="bitwise-or">|</option>
            </optgroup>
        </select>
    `;

    oCell3.innerHTML = bool2Options;

    oCell4.innerHTML = `
        <select type=text name=LogicOp2_` + predCount + `[] id=LogicOp2_` + predCount + `>
            <option value="none">None</option>

            <optgroup label="Logical Operators">
                    <option value="logic-and">&&</option>
                    <option value="logic-or">||</option>
            </optgroup>
            <optgroup label="Bitwise Operators">
                    <option value="bitwise-and">&</option>
                    <option value="bitwise-or">|</option>
            </optgroup>
        </select>
    `;

    oCell5.innerHTML = "<th><button type=button class='fa fa-trash button' style='font-size:24px;' onClick='delPredRow()'></button></th>";

    // Force the user to select a logical operator to link the next predicate
    if (curRowIndex > 1) {

        var table = document.getElementById('logic_table');
        var secondToLastRow = table.rows[curRowIndex - 1];

        // Remove None option while maintaining same id as before and remove
        secondToLastRow.cells[3].innerHTML = `
            ` + secondToLastRow.cells[3].innerHTML.split('\n')[1] + `
            <optgroup label="Logical Operators">
                    <option value="logic-and">&amp;&amp;</option>
                    <option value="logic-or">||</option>
            </optgroup>
            <optgroup label="Bitwise Operators">
                    <option value="bitwise-and">&amp;</option>
                    <option value="bitwise-or">|</option>
            </optgroup>
        </select>
                
        `
    }

}

// Deletes a logical predicate statement
function delPredRow() {

    var logicTable = document.getElementById("logic_table");
    let index = logicTable.clickedRowIndex;
    var logicRows = logicTable.rows;

    // When deleting the last logical predicate
    if (index == logicRows.length-1 && index != 1){
        console.log('deleting the last row');

        // Delete last row
        logicTable.deleteRow(document.getElementById("logic_table").clickedRowIndex);

        // Add and set None to logical operator 2 for the new last predicate using DOM manipulation
        logicRows[logicRows.length-1].cells[3].innerHTML = `
        ` + logicRows[logicRows.length-1].cells[3].innerHTML.split('\n')[1] + `
                <option value="none">None</option>
    
                <optgroup label="Logical Operators">
                        <option value="logic-and">&&</option>
                        <option value="logic-or">||</option>
                </optgroup>
                <optgroup label="Bitwise Operators">
                        <option value="bitwise-and">&</option>
                        <option value="bitwise-or">|</option>
                </optgroup>
            </select>
        `;
    } else {
        // Delete row
        logicTable.deleteRow(document.getElementById("logic_table").clickedRowIndex);
    }


}
